src/main/java/org/wildfly/security/auth/client/MatchHostRule.java

```diff
@@ -31,23 +33,46 @@
 class MatchHostRule extends MatchRule {
 
     private static final Pattern validHostSpecPattern = Pattern.compile("(?i:[-a-z0-9_]+(?:\\.[-a-z0-9_]+)*\\.?)");
+    static final Pattern validHostSpecIPv6Pattern = Pattern.compile("(?i:(?:\\:\\:)?([a-f0-9]{1,4}(?:\\:\\:?[a-f0-9]{1,4})*" +
+        "(?:\\:\\:?(?:[0-9]+(?:\\.[0-9]+)*\\.?)?)?)?)");
 
     private final String hostSpec;
 
     MatchHostRule(final MatchRule parent, final String hostSpec) {
         super(parent);
-        if (! validHostSpecPattern.matcher(hostSpec).matches()) {
-            throw log.invalidHostSpec(hostSpec);
-        }
-        if (hostSpec.endsWith(".")) {
-            this.hostSpec = hostSpec.substring(0, hostSpec.length() - 1).toLowerCase(Locale.ENGLISH);
-        } else {
-            this.hostSpec = hostSpec.toLowerCase(Locale.ENGLISH);
+        if (hostSpec.contains(":")) { // IPv6
+            if (!validHostSpecIPv6Pattern.matcher(hostSpec).matches()) {
+                throw log.invalidHostSpec(hostSpec);
+            }
+            try {
+                this.hostSpec = canonizeIPv6(hostSpec).toLowerCase(Locale.ENGLISH);
```

Comment:
As an aside, we should be using Locale.ROOT for any case-mapping in situations like these.